rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,221 @@
+- [DNS Pruning](./0008-leaderless-dns-assets/leaderless-dns-pruning.jpg)
+- [DNS Controller](./0008-leaderless-dns-assets/leaderless-dns-dns-operator.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster
+In a single cluster scenario, create a DNS Policy that references a DNS provider. When a matching host on a gateway is found to be in use, the IP addresses of that gateway will be created as A records for that host in the Zone file following the configuration outlined in the DNS Policy.
+
+### Single Cluster Health Checks
+If health checks are defined in the DNS Policy, these will also be created and used to remove unhealthy IPs from the Zone file, should all health checks fail, then all A records will be published to avoid an NXDOMAIN response.
+
+## Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret that specifies a unique `clusterID`, they should all reference the same zone(s) in the same DNS Provider for the same host. In each cluster, when a matching host on a gateway is found to be in use, the IP addresses of that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways IP addresses to the same zone file, without destroying each other's values.
+
+### Multiple Cluster Health Checks
+If health checks are defined in the DNS Policy, these will be created on all clusters, there are then 2 types of health checks:
+1) Local Health Checks - used to check this clusters IP addresses are responding for the expected host.
+2) Remote Health Checks - used to check that the other IP addresses in the zone file are responding for that zones host.
+
+When a local health check is failing, this IP will not be published to the zone.
+
+When a remote health check is failing a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
+
+If all local and remote health checks for a particular host are marked as unhealthy, all known IPs are published, to avoid an NXDOMAIN response.
+
+## OCM / ACM
+This is very similar to a single cluster, with the caveat that the "single" cluster is the hub cluster for this OCM. The primary difference is that the gateway will be a Kuadrant gatewayClass.
+
+### OCM / ACM Health Checks
+This is also the same as single cluster.
+
+## Migrating from Single cluster to Multiple cluster
+
+In the single cluster: 
+- Edit all relevant provider secrets and set:
+  - `ClusterID` to some unique value
+
+In the additional clusters:
+- Copy the DNS Policies and provider secrets from the single cluster to any additional clusters, and update the provider secrets with unique `ClusterID` values.
```

Comment:
initially I was a bit worried about this actually and if it would cause issues for git ops. But I think it is fine https://argo-cd.readthedocs.io/en/stable/operator-manual/applicationset/Generators-Cluster/ more of a FYI than anything